### PR TITLE
add guide for migrating from legacy to container infra

### PR DESCRIPTION
**Pull request by carlad**
This explain how to migrate from legacy to container-based infra.

It includes sections on:
- The benefits
- How to install apt sources and packages 
- How to install custom software.

### CODE REVIEW

**Comment by henrikhodne** [comment]
Looks good! :+1:

File: user/migrating-from-legacy.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+### Caching available for open source projects
+
+The best news for open source projects is that our build caching is now available for them too. That means faster build speeds by caching dependencies. Make sure to read the docs before trying it out.
+
+For Ruby projects, it's as simple as adding `cache: bundler` to your .travis.yml.
+
+## How can I use it?
+
+Using our new container-based stack only requires one additional line in your .travis.yml:
```
Comment (joshk) [inline user/migrating-from-legacy.md]: maybe add a note that if they see a log line about being on the container based setup then nothing needs to be done?

**Comment by joshk** [comment]
could we also add a section at the bottom which is called `Need help?` and then tell people to email support or create a github issue?

File: user/migrating-from-legacy.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+      - export PATH=$PATH:$PWD/casperjs-1.0.2/bin/
+
+Note that when you're updating the `$PATH` environment variable, that part can't be moved into a shell script, as it will only update the variable for the sub-process that's running the script.
+
+To install something from source, you can follow similar steps. Here's an example to download, compile and install the protobufs library.
+
+    install:
+      - wget https://protobuf.googlecode.com/files/protobuf-2.4.1.tar.gz
+      - tar -xzvf protobuf-2.4.1.tar.gz
+      - cd protobuf-2.4.1 && ./configure --prefix=/usr && make && sudo make install
```
Comment (joshk) [inline user/migrating-from-legacy.md]: we might want to remove sudo from this

File: user/migrating-from-legacy.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+These three commands can be extracted into a shell script, let's name it `install-protobuf.sh`:
+
+    #!/bin/sh
+    set -ex
+    wget https://protobuf.googlecode.com/files/protobuf-2.4.1.tar.gz
+    tar -xzvf protobuf-2.4.1.tar.gz
+    cd protobuf-2.4.1 && ./configure --prefix=/usr && make && sudo make install
```
Comment (joshk) [inline user/migrating-from-legacy.md]: and some more sudo

**Comment by joshk** [comment]
my very last comment, and it would be an awesome addition, would be a note on how to cache installed software.
Also, after talking to Henrik, could you create a test repo to demonstrate installing custom software, like protobuf, and then we confirm that our instructions are correct, and then we can play around with caching too.

**Comment by carlad** [comment]
Thanks for the comments @joshk 
Can you point me to where the documentation is about using cache installed software and I can add it.
I will talk to @henrikhodne about what the test repo needs for more details.

File: user/migrating-from-legacy.md (hunk 1 above)
Comment (carlad) [inline user/migrating-from-legacy.md]: done

File: user/migrating-from-legacy.md (hunk 2 above)
Comment (carlad) [inline user/migrating-from-legacy.md]: done

File: user/migrating-from-legacy.md (hunk 3 above)
Comment (carlad) [inline user/migrating-from-legacy.md]: done

**Comment by carlad** [comment]
@joshk @henrikhodne @sven I've done all the above. 
You can see the example repo here: https://github.com/travis-ci/container-example
And the builds here: https://travis-ci.org/travis-ci/container-example

Comments please.

**Comment by joshk** [comment]
Awesome work!
@roidrage are you happy with the use of 'legacy' to describe our, well, legacy setup?

**Comment by carlad** [comment]
@joshk wonder if we should add to doc the necessity for the install-protobuf.sh file needs to be chmod -x ?

**Comment by joshk** [comment]
well, either that or you can use `bash install-protobuf.sh`

**Comment by carlad** [comment]
so that would be:
`before_install:
  - bash install-protobuf.sh`
?

**Comment by joshk** [comment]
yep, that would work perfectly

**Comment by carlad** [comment]
Okay @joshk I've updated the repo and the docs. All we need now is to whether @roidrage is happy with use of the term 'legacy' to describe the legacy setup.

**Comment by carlad** [comment]
@joshk I spoke to Mathias and he has no problem with using the word 'legacy', so I think this is good to merge.

**Comment by joshk** [comment]
Fantastic. I've asked @plaindocs to give this a once over, then we can merge it in.

**Pull request merged**